xdsl/backend/riscv/register_rvscf_regalloc.py

```diff
@@ -0,0 +1,201 @@
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import MLContext, Region, SSAValue
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+@dataclass
+class RegAllocCtx:
+    """
+    This is the context object for allocating registers.
+
+    It keeps track of how many times each ssa value is expected to be seen before
+    it is "free" again.
+
+    It also carries a list of forbidden registers, as they are used by the "outside"
+    (i.e. parent region) and shall not be overwritten.
+
+    This is currently still a bit broken, as ssa values can have the same register
+    even before.
+    """
+
+    liveliness: dict[riscv.RISCVRegisterType, int] = field(
+        default_factory=lambda: defaultdict(int)
+    )
+    """
+    dictionary mapping registers to how many uses they have "left"
+    """
+
+    forbidden_vals: set[riscv.RISCVRegisterType] = field(default_factory=set)
+    """
+    A set of registers that are forbidden for allocation
+    """
+
+    def child_for_loops(self):
+        """
+        Create a new context that cannot overwrite any registers from the "parent"
+        context.
+
+        Moves all registers with remaining uses to forbidden_vals
+        """
+        return RegAllocCtx(
+            forbidden_vals={reg for reg, cnt in self.liveliness.items() if cnt > 0}
+        )
+
+    def free_int_reg(self) -> riscv.IntRegisterType:
+        for reg in (
+            *riscv.Registers.A,
+            *riscv.Registers.T,
+            *riscv.Registers.S,
+        ):
+            if reg not in self.forbidden_vals and self.liveliness[reg] == 0:
+                return reg
+        raise NotImplementedError("Out of int registers, spilling not implemented")
+
+    def free_float_reg(self) -> riscv.FloatRegisterType:
+        for reg in (
+            *riscv.Registers.FA,
+            *riscv.Registers.FT,
+            *riscv.Registers.FS,
+        ):
+            if reg not in self.forbidden_vals and self.liveliness[reg] == 0:
+                return reg
+        raise NotImplementedError("Out of float registers, spilling not implemented")
+
+    def add_reg(self, val: SSAValue):
+        assert isinstance(val.type, riscv.RISCVRegisterType) and val.type.is_allocated
+        if self.liveliness[val.type] > 0:
+            # this may be interesting behaviour? idk
+            pass
+        self.liveliness[val.type] = len(val.uses)
+
+    def register_use(self, val: SSAValue):
+        assert isinstance(val.type, riscv.RISCVRegisterType) and val.type.is_allocated
+        if self.liveliness[val.type] > 0:
+            self.liveliness[val.type] -= 1
+
+    def make_forbidden(self, reg: riscv.RISCVRegisterType):
+        self.forbidden_vals.add(reg)
+
+
+def register_allocate_function(func: riscv_func.FuncOp):
+    # create an empty context
+    ctx = RegAllocCtx()
```

Comment:
Maybe irrelevant right now, but will/should this still work for nested functions?